coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -225,11 +225,89 @@ class SingleClusterPlanner(dsRef: DatasetRef,
       case lp: ScalarFixedDoublePlan       => materializeFixedScalar(qContext, lp)
       case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp)
       case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp)
+      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp)
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
       case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
+  /**
+   * Example1:
+   * min_over_time(<someQueryReturningInstantVector>[3m:1m])
+   *
+   * outerStart = S
+   * outerEnd = E
+   * outerStep = 90s
+   *
+   * Resulting Exec Plan:
+   *
+   * - ApplySubqueryRangeFunction MinOverTime from S to E, at lookback of 3m, innerStep of 60s, outerStep = 90s
+   *     - Execute Range Query someQueryReturningInstantVector from start=S-3m until end=E at step=30 GCD(60,90)
+   *
+   *
+   * Example2:
+   * min_over_time(avg_over_time(<someNestedQueryReturningInstantVector>[6m:2m])[3m:1m])
+   *
+   * outerStart = S
+   * outerEnd = E
+   * outerStep = 90s
+   *
+   * - ApplySubqueryRangeFunction MinOverTime from S to E, at lookback of 3m, innerStep of 60s, outerStep = 90s
+   *    - Execute Range Query sum(avg_over_time(<anotherNestedQueryReturningInstantVector>[6m:2m]))
+   *      from start=S-3m until end=E at step=30s i.e. GCD(60s,90s)
+   *         - ApplySubqueryRangeFunction AvgOverTime from S-3m to E,
+   *           at lookback of 6m, innerStep of 120s, outerStep = 30s
+   *                - Execute Range Query <anotherNestedQueryReturningInstantVector>
+   *                  from start=S-3m-6m until end=E at step=30s i.e. GCD(30s,120s)
+   */
+  private def materializeSubqueryWithWindowing(qContext: QueryContext, sqww: SubqueryWithWindowing): PlanResult = {
+    // This physical plan will try to execute only one range query instead of a number of individual subqueries.
+    // If ranges of each of the subqueries have overlap, retrieving the total range
+    // is optimal, if there is no overlap and even worse significant gap between the individual subqueries, retrieving
+    // the entire range might be suboptimal, this still might be a better option than issuing and concatenating numerous
+    // subqueries separately
+    val rangeFn = InternalRangeFunction.lpToInternalFunc(sqww.functionId)
+    var innerPeriodicSeriesPlan = sqww.innerPeriodicSeries
+    val rangeVectorTransformer = if (queryConfig.fastSubquery) {
+      val paramsExec = materializeFunctionArgs(sqww.functionArgs, qContext)
+      val window = if (rangeFn == InternalRangeFunction.Timestamp)
+        None
```

Comment:
Why is window None when timestamp function is used? Deserves a comment to help reader understand the reason since it is non-obvious.

Reply:
actually good catch, technically, timestamp CANNOT be a function for a subquery argument as it only accepts instant vectors. The reason it was here, is because I was creatively copying PeriodicSeriesWithWindowing code and it had this piece of code there. For PeriodicSeriesWithWindowing it makes sense, not so much for SubqueryWithWindowing. Will remove the logic.